### ISSUE TITLE
EVP_PKEY_get_group_name(3) wrong suggested buffer size or misleading docs.

### ISSUE THREAD

**Issue by vdukhovni** [openssl/openssl#16287]
Migrating some code that logs EC group names to the new 3.0 (non-deprecated) API, I had to increment the reported required length to get a buffer large enough to accomodate the trailing `NUL`.  Otherwise, the returned string was not NUL-terminated.
```
    size_t  namelen;

    if (EVP_PKEY_get_group_name(pkey, 0, 0, &namelen)) {
        curve = malloc(++namelen);
        if (!EVP_PKEY_get_group_name(pkey, curve, namelen, 0)) {
            free(curve);
            curve = 0;
        }
    }
```
The documentation says:
```
    EVP_PKEY_get_group_name() fills in the group name of the *pkey* into
    *gname*, up to at most *gname_sz* bytes including the ending NUL byte
    and assigns **gname_len* the actual size of the name, if *pkey*'s key
    type supports it. *gname* as well as *gname_len* may individually be
    NULL, and won't be filled in or assigned in that case.
```
So on input `gname_sz` includes the NUL byte, which leads one to speculate that the output `gname_len` also counts the NUL byte (size not string length of name), but this appears to not be the case (at least for ECDHE keys, where "orime256v1" reports 10 bytes needed, but in fact we need 11 for the NUL byte.

**Comment by t8m** [comment]
IMO this is by design but it might be the documentation needs clarification. The function calls EVP_PKEY_get_utf8_string_param() under the hood and the parameters are just forwarded to that function. IMO the intention of the `assigns **gname_len* the actual size of the name` is that the size of the name is length of the string not including the NUL byte. The NUL termination is something that is added only if it fits the gname_size and that is also by design.

Any suggestions on how to clarify the documentation would be welcome.

**Comment by vdukhovni** [comment]
Make it explicit that the returned `gname_len` is "the length of the string *excluding* the terminating NUL byte" rather than the current "the actual size of the name" (when I read "size", I usually think buffer size not string length, and the context also slants in that direction).

**Comment by t8m** [comment]
Fix in #16296 in addition to other numerous documentation bug fixes which I've found alongside. So thanks @vdukhovni for reporting this issue to make me review this mess. :grin:
